hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/parse/QueryParser.java

```diff
@@ -19,100 +19,70 @@
 import com.hazelcast.sql.SqlErrorCode;
 import com.hazelcast.sql.impl.JetSqlBackend;
 import com.hazelcast.sql.impl.QueryException;
-import com.hazelcast.sql.impl.calcite.parser.HazelcastSqlParser;
 import com.hazelcast.sql.impl.calcite.validate.HazelcastSqlConformance;
-import org.apache.calcite.rel.type.RelDataType;
-import org.apache.calcite.sql.SqlCall;
-import org.apache.calcite.sql.SqlHint;
 import org.apache.calcite.sql.SqlKind;
 import org.apache.calcite.sql.SqlNode;
 import org.apache.calcite.sql.SqlOperator;
-import org.apache.calcite.sql.SqlOperatorTable;
-import org.apache.calcite.sql.SqlSelect;
 import org.apache.calcite.sql.parser.SqlParser;
-import org.apache.calcite.sql.util.SqlBasicVisitor;
+import org.apache.calcite.sql.parser.SqlParserImplFactory;
 import org.apache.calcite.sql.validate.SqlValidator;
 
-import java.util.Collections;
-import java.util.HashSet;
+import javax.annotation.Nullable;
 import java.util.Set;
 
+import static java.util.Collections.emptySet;
+
 /**
  * Performs syntactic and semantic validation of the query, and converts the parse tree into a relational tree.
  */
 public class QueryParser {
 
-    /** A hint to force execution of a query on Jet. */
-    private static final String RUN_ON_JET_HINT = "jet";
+    private final SqlParser.Config parserConfig;
 
-    private static final SqlParser.Config CONFIG;
     private final SqlValidator validator;
-
-    static {
+    private final Set<SqlKind> extensionSqlKinds;
+    private final Set<SqlOperator> extensionSqlOperators;
+
+    @SuppressWarnings("unchecked")
+    public QueryParser(
+            SqlValidator validator,
+            @Nullable JetSqlBackend jetSqlBackend
+    ) {
         SqlParser.ConfigBuilder configBuilder = SqlParser.configBuilder();
-
         CasingConfiguration.DEFAULT.toParserConfig(configBuilder);
         configBuilder.setConformance(HazelcastSqlConformance.INSTANCE);
-        configBuilder.setParserFactory(HazelcastSqlParser.FACTORY);
-
-        CONFIG = configBuilder.build();
-    }
+        if (jetSqlBackend != null) {
+            configBuilder.setParserFactory((SqlParserImplFactory) jetSqlBackend.createParserFactory());
+        }
+        this.parserConfig = configBuilder.build();
 
-    public QueryParser(SqlValidator validator) {
         this.validator = validator;
+        this.extensionSqlKinds = jetSqlBackend == null ? emptySet() : (Set<SqlKind>) jetSqlBackend.kinds();
```

Comment:
Changes in this class look suspicious to me - what if we reveal that some context information that is needed for a new command is not passed here? 
An alternative is to move the whole parser instantiation to `jetSqlBackend`. This way the risk of the aforementioned problem is zero.

Reply:
I have slightly refactored `QueryParser`, `UnsupportedOperationVisitor` is specific to either IMDG & JET